Animation longhand serialization should be a list of each animation rather than a list of lists for each longhand.

absoludity opened the pull request:
Fixes part of #15398 -  serialization should be a list of each animation, rather than a list of lists for each longhand.

---
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes are prework for #15398

- [X] There are tests for these changes OR
- [ ] These changes do not require tests because _____

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/15516)
<!-- Reviewable:end -->

highfive commented:
Thanks for the pull request, and welcome! The Servo team is excited to review your changes, and you should hear from @jdm (or someone else) soon.

highfive commented:
Heads up! This PR modifies the following files:
 * @bholley: components/style/properties/longhand/box.mako.rs
 * @emilio: components/style/properties/longhand/box.mako.rs

jdm commented:
r? @Manishearth

Manishearth commented on tests/unit/style/properties/serialization.rs:
```diff
@@ -1114,4 +1114,85 @@ mod shorthand_serialization {
+        use style::properties::longhands::animation_timing_function;
+        use style::values::specified::Time as TimeContainer;
+
+        macro_rules! create_animation_properties {
```
You probably should be able to specify a vector of each property, since we're testing for cases where the vectors don't match.

Manishearth commented:
Looking good so far.

Manishearth commented:
It's also okay to parse a PropertyDeclarationBlock from CSS source for each test and then attempt to extract the animation shorthand. This may make it easier to write tests (since you just need to provide CSS source), and will reduce the size of the code.

absoludity commented:
Great, thanks @Manishearth . I'll parse a PropertyDeclarationBlock instead of the macro - that will be a simpler way to test both the single and multiple animation case. Thanks.

absoludity committed:
Fix serialization of animation.

absoludity commented:
@Manishearth I've updated the tests to parse a PropertyDeclarationBlock instead, which is much nicer - thanks.

I also updated the code which was mentioned in the issue: shorthand/box.mako.rs so that the tests work as expected for the single and multiple animation.

Where things don't work (I think?) is the empty condition you mention:

> For computation, the value should be cycled, but for serialization of specified value, it should just return empty if the list lengths don't match each other.

As you can see from the test, by modifying box.mako.rs to catch the unequal vector lengths and return "", the serialized empty result is then "animation: ;", which I assume is *not* what is wanted here? Should I instead catch that condition in PropertyDeclarationBlock.to_css , or where abouts? Let me know.

Thanks again for the help.

Manishearth commented on tests/unit/style/properties/serialization.rs:
```diff
@@ -1114,4 +1114,107 @@ mod shorthand_serialization {
+
+            let serialization = block.to_css_string();
+
+            assert_eq!(serialization, "animation: ;");
```
This shouldn't serialize as a shorthand, no?

absoludity replied:
@Manishearth Yes, I think so - that's what my question in the comment above where I'm asking for you feedback or thoughts, is all about :)

I didn't push this to land, but just to get your thoughts on my comment above, when you have time. Thanks.

Manishearth replied:
Yeah, the tests look fine. They're testing erroneous behavior, but aside from that fine. Are you going to fix the erroneous behavior in this same pull request (better if you do, but if you want to hold that off, mark the erroneous test cases with a comment pointing to #15398)

(It helps to test in Firefox in Javascript with element.style.animation)

absoludity replied:
@Manishearth Sorry, my question on which I was after feedback above was:

> As you can see from the test, by modifying box.mako.rs to catch the unequal vector lengths and return "", the serialized empty result is then "animation: ;", which I assume is not what is wanted here? Should I instead catch that condition in PropertyDeclarationBlock.to_css , or where abouts? Let me know.

So yep, I was aware of and am planning on fixing the erroneous behavior, but wasn't sure if PropertyDeclarationBlock.to_css would be the correct place (the original issue just pointed at box.mako.rs which doesn't seem sufficient).

Manishearth replied:
The correct place is to fix box.mako.rs I think? You need to make it return an error in these cases. cc @upsuper

upsuper replied:
Should it return an error? Would that causes the serialization to fail completely? It makes me think that Servo's current mechanism probably cannot handle this kind of cases that all longhands are specified but the shorthand is not serializable. This issue would show again in #15036.

absoludity replied:
@upsuper @Manishearth OK, in that case, is it OK if I reduce the scope of my branch to fix the main issue on the bug only - that the serialization of multiple animations is correct, and treat the other case of unequal value list lengths separately?

Manishearth replied:
Yeah, that probably will cause it to fail. Perhaps we should reinterpret the Err in shorthand serializers. But yes, feel free to reduce the scope.

Wafflespeanut commented on components/style/properties/shorthand/box.mako.rs:
```diff
@@ -308,28 +308,88 @@ macro_rules! try_parse_one {
 
-            try!(self.animation_iteration_count.to_css(dest));
-            try!(write!(dest, " "));
+            // usuper says on #15398: For computation, the value should be cycled, but for
```
Nit: Please change this comment to something like,
`// serialize only when all the lists have same length`

Wafflespeanut commented on components/style/properties/shorthand/box.mako.rs:
```diff
@@ -308,28 +308,88 @@ macro_rules! try_parse_one {
+                    try!(write!(dest, ", "));
+                }
+
+                if let Some(duration) = duration {
```
Why couldn't this (and the following) be generated by Mako code?

absoludity replied:
Thanks - much neater.

Wafflespeanut commented:
./components/style/properties/shorthand/box.mako.rs:337: Line is longer than 120 characters
    ./tests/unit/style/properties/serialization.rs:1165: Line is longer than 120 characters

bors-servo commented:
:umbrella: The latest upstream changes (presumably #15546) made this pull request unmergeable. Please resolve the merge conflicts.

absoludity commented:
OK, I've updated this so that it no longer tries to tackle:

> animation shorthand should be serializable only when all the lists have the same length.

but does fix:

> And the serialization should be a list of each animation, rather than a list of lists for each longhand.

That way it should be landable as is. I've noted the issue of the incorrect behaviour (cycling of item values if lists have unequal lengths), and would be happy to try to tackle that next. Thanks.

Manishearth commented:
r? @upsuper

absoludity commented:
@upsuper If you get a chance to review/land this, that'd be super.

I took a quick look at whether I can also address the first part of the issue, that

> animation shorthand should be serializable only when all the lists have the same length.

and it's certainly possible from a quick play by just handling the (currently unhandled) error in ToCss::to_css_string like:

```
diff --git a/components/style_traits/values.rs b/components/style_traits/values.rs
index ab73050..dd18078 100644
--- a/components/style_traits/values.rs
+++ b/components/style_traits/values.rs
@@ -19,8 +19,10 @@ pub trait ToCss {
     #[inline]
     fn to_css_string(&self) -> String {
         let mut s = String::new();
-        self.to_css(&mut s).unwrap();
-        s
+        match self.to_css(&mut s) {
+            Ok(()) => s,
+            Err(fmt::Error) => String::new(),
+        }
     }
 }
```

after returning an Err(fmt::Error) in the appropriate place in LonghandToSerialize::to_css_declared. Let me know if that's not the right place (maybe an implementation closer to Animation could override it), and I'll do a second branch which handles this too (and can update Background and Transition too - if the direction is correct).

upsuper commented on tests/unit/style/properties/serialization.rs:
```diff
@@ -1136,6 +1136,117 @@ mod shorthand_serialization {
+                animation-direction: ;\
+                animation-fill-mode: ;\
+                animation-iteration-count: ;\
+                animation-play-state: ;");
```
This doesn't make much sense, since all the declarations here are just invalid... You should probably try having all but one longhands specified, and test that the generated string is still empty whatever the one unspecified longhand is.

absoludity replied:
Yes - I only added it because I saw the code writing "" if the (max) length of the attribute values was zero. Removing.

upsuper replied:
Oh, wait, the string wouldn't be empty if there is any property... It's probably fine to just remove this test...

upsuper commented on components/style/properties/shorthand/box.mako.rs:
```diff
@@ -308,28 +308,55 @@ macro_rules! try_parse_one {
-            try!(write!(dest, " "));
+            use std::cmp;
+            let mut len = 0;
+            % for name in "duration timing_function delay direction fill_mode iteration_count play_state name".split():
```
Please leave a TODO comment here saying that when their length are different, shorthand should not be serialized at all.

absoludity replied:
Done. Thanks @upsuper for the review and feedback - I'll take a stab at the other issue over the weekend.

upsuper replied:
Thanks for working on this.

upsuper reviewed (commented):
I'm probably fine with this change given the lack of more fundamental mechanism to stop shorthand from being generated. We can fix that latter.

I'll r+ with the following comments addressed.

upsuper commented:
> after returning an Err(fmt::Error) in the appropriate place in LonghandToSerialize::to_css_declared. Let me know if that's not the right place (maybe an implementation closer to Animation could override it), and I'll do a second branch which handles this too (and can update Background and Transition too - if the direction is correct).

I don't think that is the correct place. The hard part is that, we need to report to the algorithm of `PropertyDeclarationBlock::to_css()` that we fail to serialize the shorthand, but not because of IO error, so that the corresponding longhands can be serialized in themselves.

upsuper commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 368f459 has been approved by `upsuper`

<!-- @bors-servo r=upsuper 368f4591530f173d72e919f616de0d8c2e0bc4b2 -->

bors-servo commented:
:hourglass: Testing commit 368f459 with merge 2ef216f...

bors-servo commented:
:broken_heart: Test failed - [windows-gnu-dev](http://build.servo.org/builders/windows-gnu-dev/builds/874)

upsuper commented:
> CompileException: Fragment 'for name in "duration timing_function delay direction fill_mode \' is not a partial control statement in file 'C:/buildbot/slave/windows-gnu-dev/build/components/style/properties/shorthand/box.mako.rs' at line: 334 char: 1

You probably cannot break line like that.

absoludity commented:
@upsuper Hrm - it works fine on linux and seems to be supported from the docs , but anyway, I've out-dented the make tags (which seems to make sense - so the rust code is correctly indented in the resulting rs file) so that test-tidy passes (the reason for the line break was test-tidy complaining about long lines). Not sure if test-tidy should enforce the 120c line limit on mako syntax? Anyway, let me know if the outdent is OK.

absoludity commented:
@upsuper Nudge, just before it needs another rebase :)

upsuper commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 75a993c has been approved by `upsuper`

<!-- @bors-servo r=upsuper 75a993ce0cad444d2f716970aa2b97eb0f6bb978 -->

upsuper commented:
Sorry for the delay. I didn't notice the new commit.

bors-servo commented:
:hourglass: Testing commit 75a993c with merge 978a603...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/5440), [arm32](http://build.servo.org/builders/arm32/builds/5651), [arm64](http://build.servo.org/builders/arm64/builds/5680), [linux-dev](http://build.servo.org/builders/linux-dev/builds/5792), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/2485), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/2506), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/5781), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/5811), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/2464), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/2529), [windows-gnu-dev](http://build.servo.org/builders/windows-gnu-dev/builds/941), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/943)
Approved by: upsuper
Pushing 978a603190dc7230c95ef9ecaca9dcdb1a78bf4c to master...
